Flyway migration problem

## Situation
Having a database with a running SecHub server v0.28.0, I wanted to update to Server v0.29.0 the server start did fail because of a FlywayException

## Wanted
Server shall start without problems

## Solution
Open

Here the stacktrace:
```
2022-03-03 08:41:48.500 ERROR 19869 --- [           main] o.s.boot.SpringApplication               : Application run failed

org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'flywayInitializer' defined in class path resource [org/springframework/boot/autoconfigure/flyway/FlywayAutoConfiguration$FlywayConfiguration.class]: Invocation of init method failed; nested exception is org.flywaydb.core.api.exception.FlywayValidateException: Validate failed: Migrations have failed validation
Migration checksum mismatch for migration version 1
-> Applied to database : 715094934
-> Resolved locally    : 1869507521. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 2
-> Applied to database : -1701083331
-> Resolved locally    : -1915064674. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 3
-> Applied to database : -18473243
-> Resolved locally    : -1360325796. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 4
-> Applied to database : -189469386
-> Resolved locally    : 559734107. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 5
-> Applied to database : 1350730081
-> Resolved locally    : -341752538. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 6
-> Applied to database : -948723580
-> Resolved locally    : -795912523. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 7
-> Applied to database : -1520585304
-> Resolved locally    : -1838403430. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 8
-> Applied to database : 1734111914
-> Resolved locally    : -1939194000. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 9
-> Applied to database : 921741995
-> Resolved locally    : 279988134. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 10
-> Applied to database : -1984759970
-> Resolved locally    : 1965935317. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 11
-> Applied to database : -461865474
-> Resolved locally    : 1301434468. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 12
-> Applied to database : 1454304337
-> Resolved locally    : 2143662752. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 13
-> Applied to database : 546297426
-> Resolved locally    : -2040436880. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 14
-> Applied to database : 1703978141
-> Resolved locally    : 480783958. Either revert the changes to the migration, or run repair to update the schema history.
Need more flexibility with validation rules? Learn more: https://rd.gt/3AbJUZE
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1804)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:620)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:322)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)
	at org.springframework.context.support.AbstractApplicationContext.getBean(AbstractApplicationContext.java:1154)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:908)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:583)
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:145)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:730)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:412)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:302)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1301)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1290)
	at com.mercedesbenz.sechub.SecHubServerApplication.main(SecHubServerApplication.java:26)
Caused by: org.flywaydb.core.api.exception.FlywayValidateException: Validate failed: Migrations have failed validation
Migration checksum mismatch for migration version 1
-> Applied to database : 715094934
-> Resolved locally    : 1869507521. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 2
-> Applied to database : -1701083331
-> Resolved locally    : -1915064674. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 3
-> Applied to database : -18473243
-> Resolved locally    : -1360325796. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 4
-> Applied to database : -189469386
-> Resolved locally    : 559734107. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 5
-> Applied to database : 1350730081
-> Resolved locally    : -341752538. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 6
-> Applied to database : -948723580
-> Resolved locally    : -795912523. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 7
-> Applied to database : -1520585304
-> Resolved locally    : -1838403430. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 8
-> Applied to database : 1734111914
-> Resolved locally    : -1939194000. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 9
-> Applied to database : 921741995
-> Resolved locally    : 279988134. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 10
-> Applied to database : -1984759970
-> Resolved locally    : 1965935317. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 11
-> Applied to database : -461865474
-> Resolved locally    : 1301434468. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 12
-> Applied to database : 1454304337
-> Resolved locally    : 2143662752. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 13
-> Applied to database : 546297426
-> Resolved locally    : -2040436880. Either revert the changes to the migration, or run repair to update the schema history.
Migration checksum mismatch for migration version 14
-> Applied to database : 1703978141
-> Resolved locally    : 480783958. Either revert the changes to the migration, or run repair to update the schema history.
Need more flexibility with validation rules? Learn more: https://rd.gt/3AbJUZE
	at org.flywaydb.core.Flyway$1.execute(Flyway.java:131)
	at org.flywaydb.core.Flyway$1.execute(Flyway.java:124)
	at org.flywaydb.core.FlywayExecutor.execute(FlywayExecutor.java:214)
	at org.flywaydb.core.Flyway.migrate(Flyway.java:124)
	at org.springframework.boot.autoconfigure.flyway.FlywayMigrationInitializer.afterPropertiesSet(FlywayMigrationInitializer.java:66)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1863)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1800)
	... 18 common frames omitted

```

--- a/sechub-pds/src/main/java/com/mercedesbenz/sechub/pds/PDSFlywayFactory.java
+++ b/sechub-pds/src/main/java/com/mercedesbenz/sechub/pds/PDSFlywayFactory.java
@@ -11,8 +11,10 @@ import org.springframework.beans.factory.annotation.Value;
 import org.springframework.boot.autoconfigure.flyway.FlywayMigrationInitializer;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Profile;
 
 @Configuration
+@Profile("!" + PDSProfiles.TEST) // spring profile "test" has no Flyway enabled
 public class PDSFlywayFactory {
 
     private static final Logger LOG = LoggerFactory.getLogger(PDSFlywayFactory.class);

--- a/sechub-pds/src/main/java/com/mercedesbenz/sechub/pds/PDSFlywayFactory.java
+++ b/sechub-pds/src/main/java/com/mercedesbenz/sechub/pds/PDSFlywayFactory.java
@@ -1,0 +1,52 @@
+package com.mercedesbenz.sechub.pds;
+
+import java.util.List;
+
+import org.flywaydb.core.Flyway;
+import org.flywaydb.core.api.output.RepairOutput;
+import org.flywaydb.core.api.output.RepairResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.autoconfigure.flyway.FlywayMigrationInitializer;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class PDSFlywayFactory {
+
+    private static final Logger LOG = LoggerFactory.getLogger(PDSFlywayFactory.class);
+
+    @Value("${pds.migration.flyway.autorepair:true}")
+    @PDSMustBeDocumented(value = "When enabled, flyway migration problems will be automatically repaired", scope = "migration")
+    boolean repairAutomatically;
+
+    @Bean
+    public FlywayMigrationInitializer flywayInitializer(Flyway flyway) {
+        if (repairAutomatically) {
+            LOG.info("Start flyway repair");
+
+            RepairResult result = flyway.repair();
+
+            logRepairs("align migration", result.migrationsAligned);
+            logRepairs("delete migration", result.migrationsDeleted);
+            logRepairs("remove migration", result.migrationsRemoved);
+
+            LOG.info("Flyway repair done for flyway version:{}", result.flywayVersion);
+
+        }
+        return new FlywayMigrationInitializer(flyway);
+    }
+
+    private void logRepairs(String message, List<RepairOutput> outputList) {
+        for (RepairOutput output : outputList) {
+            logRepair(message, output);
+        }
+
+    }
+
+    private void logRepair(String repairAction, RepairOutput output) {
+        LOG.info("Flyway repair did {}: version={},description={},", repairAction, output.version, output.description);
+    }
+
+}

--- a/sechub-server/src/main/java/com/mercedesbenz/sechub/server/SecHubServerFlywayFactory.java
+++ b/sechub-server/src/main/java/com/mercedesbenz/sechub/server/SecHubServerFlywayFactory.java
@@ -11,10 +11,13 @@ import org.springframework.beans.factory.annotation.Value;
 import org.springframework.boot.autoconfigure.flyway.FlywayMigrationInitializer;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Profile;
 
 import com.mercedesbenz.sechub.sharedkernel.MustBeDocumented;
+import com.mercedesbenz.sechub.sharedkernel.Profiles;
 
 @Configuration
+@Profile("!" + Profiles.TEST) // spring profile "test" has no Flyway enabled
 public class SecHubServerFlywayFactory {
 
     private static final Logger LOG = LoggerFactory.getLogger(SecHubServerFlywayFactory.class);

--- a/sechub-server/src/main/java/com/mercedesbenz/sechub/server/SecHubServerFlywayFactory.java
+++ b/sechub-server/src/main/java/com/mercedesbenz/sechub/server/SecHubServerFlywayFactory.java
@@ -1,0 +1,54 @@
+package com.mercedesbenz.sechub.server;
+
+import java.util.List;
+
+import org.flywaydb.core.Flyway;
+import org.flywaydb.core.api.output.RepairOutput;
+import org.flywaydb.core.api.output.RepairResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.autoconfigure.flyway.FlywayMigrationInitializer;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+import com.mercedesbenz.sechub.sharedkernel.MustBeDocumented;
+
+@Configuration
+public class SecHubServerFlywayFactory {
+
+    private static final Logger LOG = LoggerFactory.getLogger(SecHubServerFlywayFactory.class);
+
+    @Value("${sechub.migration.flyway.autorepair:true}")
+    @MustBeDocumented(value = "When enabled, flyway migration problems will be automatically repaired", scope = "migration")
+    boolean repairAutomatically;
+
+    @Bean
+    public FlywayMigrationInitializer flywayInitializer(Flyway flyway) {
+        if (repairAutomatically) {
+            LOG.info("Start flyway repair");
+
+            RepairResult result = flyway.repair();
+
+            logRepairs("align migration", result.migrationsAligned);
+            logRepairs("delete migration", result.migrationsDeleted);
+            logRepairs("remove migration", result.migrationsRemoved);
+
+            LOG.info("Flyway repair done for flyway version:{}", result.flywayVersion);
+
+        }
+        return new FlywayMigrationInitializer(flyway);
+    }
+
+    private void logRepairs(String message, List<RepairOutput> outputList) {
+        for (RepairOutput output : outputList) {
+            logRepair(message, output);
+        }
+
+    }
+
+    private void logRepair(String repairAction, RepairOutput output) {
+        LOG.info("Flyway repair did {}: version={},description={},", repairAction, output.version, output.description);
+    }
+
+}

